mypy/newsemanal/semanal.py

```diff
@@ -445,10 +445,13 @@ def add_builtin_aliases(self, tree: MypyFile) -> None:
                     del tree.names[name]
 
     def adjust_public_exports(self) -> None:
-        """Make variables not in __all__ not be public"""
+        """Make variables in __all__ be exported, and variables not in
+        __all__ not be public"""
         if '__all__' in self.globals:
             for name, g in self.globals.items():
-                if name not in self.all_exports:
+                if name in self.all_exports:
+                    g.module_hidden = False
```

Comment:
Looking at the docs for `module_public` and `module_hidden`:
 * public means that it will be picked up by an `import *`
 * hidden means that it can't be imported at all

So I think we need to set `public = True` when the name does appear to prevent it from being spurious excluded from `import *`.